appinventor/blocklyeditor/src/field_lexical_variable.js

```diff
@@ -100,19 +102,25 @@ Blockly.FieldLexicalVariable.prototype.setValue = function(text) {
  * Update the eventparam mutation associated with the field's source block.
  */
 Blockly.FieldLexicalVariable.prototype.updateMutation = function() {
+  // TODO: I think I need someone to explain this function to me.
+
   var text = this.getText();
+  // TODO: Change to if return;
   if (this.sourceBlock_ && this.sourceBlock_.getParent()) {
     this.sourceBlock_.eventparam = undefined;
+    // TODO: Is this necessary? I don't think global vars can have parents.
     if (text.indexOf(Blockly.Msg.LANG_VARIABLES_GLOBAL_PREFIX + ' ') === 0) {
       this.sourceBlock_.eventparam = null;
       return;
     }
+
     var i, parent = this.sourceBlock_.getParent();
     while (parent) {
       var variables = parent.declaredVariables ? parent.declaredVariables() : [];
       if (parent.type != 'component_event') {
         for (i = 0; i < variables.length; i++) {
           if (variables[i] == text) {
+            // TODO: Wouldn't this be tracking the outermost scope?
```

Comment:
It's the innermost scope, we're starting at a block in the stack and walking upward to the top level. Consider the following example:

```
when Camera1.AfterPicture(image) do:
  initialize local variable image = "" in:
    set Button1.Image = image
```

If we are determining whether `image` should have an `eventparam` mutation, we start at image and walk up the tree to find the closest declaration of `image`, which is the local variable definition, not the event handler definition. Therefore, we can set eventparam to null.

Reply:
Gotcha! Thank you for the example that makes this much clearer.